src/metabase/api/public.clj

```diff
@@ -531,6 +553,44 @@
      :export-format :api
      :parameters    parameters :qp-runner qp.pivot/run-pivot-query)))
 
+(def ^:private action-execution-throttle
+  "Rate limit at 1 action per second on a per action basis.
+   The goal of rate limiting should be to prevent very obvious abuse, but it should
+   be relatively lax so we don't annoy legitimate users."
+  (throttle/make-throttler :action-uuid :attempts-threshold 1 :initial-delay-ms 1000 :delay-exponent 1))
+
+#_{:clj-kondo/ignore [:deprecated-var]}
+(api/defendpoint-schema POST "/action/:uuid/execute"
+  "Execute the Action.
+
+   `parameters` should be the mapped dashboard parameters with values.
+   `extra_parameters` should be the extra, user entered parameter values."
+  [uuid :as {{:keys [parameters], :as _body} :body}]
+  {uuid        su/UUIDString
+   parameters  (s/maybe {s/Keyword s/Any})}
+  (let [throttle-message (try
+                           (throttle/check action-execution-throttle uuid)
+                           nil
+                           (catch ExceptionInfo e
+                             (get-in (ex-data e) [:errors :action-uuid])))
+        throttle-time (when throttle-message
+                        (second (re-find #"You must wait ([0-9]+) seconds" throttle-message)))]
+    (if throttle-message
+      (cond-> {:status 429
+               :body   throttle-message}
+        throttle-time (assoc :headers {"Retry-After" throttle-time}))
+      (do
+        ;; TODO: check-actions-enabled for the database
+        (validation/check-public-sharing-enabled)
+        ;; Run this query with full superuser perms. We don't want the various perms checks
+        ;; failing because there are no current user perms; if this Dashcard is public
+        ;; you're by definition allowed to run it without a perms check anyway
+        (binding [api/*current-user-permissions-set* (delay #{"/"})]
```

Comment:
shouldn't this be an `atom` instead of delay?

Reply:
I'm not sure to be honest. Both `delay` and `atom` are used throughout the codebase. But I can't see any reason to use `atom` over `delay`, since it doesn't appear to be swapped / reset anywhere.

Reply:
I think we only use `delay` in tests. 

In apis, we use `atom`, there are good amount of this pattern in `api/public`

Reply:
My understanding is that `delay` is the better choice when possible since it communicates that you're not mutating it. (And of course the `delay` is necessary at all since the value gets derefed)

Reply:
I'm not sure why the value needs to be a `delay` or `atom` at all, which is why the choice is confusing. Maybe it's for historical reasons. Could it just be a dynamic var?

Reply:
we used `delay` as an optimization for APIs don't need to query permissions.

The same for `curernt-user, user-locale` : https://github.com/metabase/metabase/blob/d5c768af1263fc98c84ff17877b7d50972a8e26d/src/metabase/server/middleware/session.clj#L312

Reply:
@tsmacdonald good reasoning. Yea I agree it's better to use `delay` here now.